Initial work for Qdrant resource and component

timheuer opened the pull request:
Initial proposal for #795

This PR adds a Resource and initial Component (plus playground and tests)

Resource exposes two API access endpoints (REST + gRPC). The QdrantClient for .NET uses gRPC by default but Semantic Kernel does not use that client library so the other exposed endpoint is helpful.

REST endpoint by default exposes a Web UI dashboard -- this is excluded in Publish (per documentation in source, confirmed with Qdrant contributors).

Endpoints are secred non-optionally with an API Key (using ParameterResource -- will generate random if not provided).

Component used QdrantClient .NET version (recommended from Qdrant). This component uses gRPC endpoint by default. Settings will read from standard env vars for ConnectionString + API key or from config for the component itself if provided.

The component doesn't seem to be rich in trace/metrics so those are not implemented...ideas?

Little runtime demo with playground+dashboard apps:


https://github.com/dotnet/aspire/assets/4821/eec32fbc-7096-4485-bdbb-27441f76a094


 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3131)

timheuer committed:
Initial draft of QdrantServerResource
- exposed gRPC endpoint as primary (c# client uses that)
- exposes HTTP endpoint optionally (as dashboard is there)
- defaults to secure the access via API key (mainly due to dashboard)

timheuer committed:
Fix-up after main rebase on removing InputAnnotation

timheuer committed:
Remove dashboard in Publish mode
- change how dashboard is removed in Publish (per docs)
- add 'qdrant' to spelling.dic
- fix tests

timheuer committed:
Fixup code style violations

timheuer committed:
Change ApiKey reference
- Add WithReference overload (setting cn string + key)
- Added tests for named parameter on manifest
- Changed playground app

timheuer committed:
Initial Qdrant component work
- adds component (using Qdrant.QdrantClient)
- Change playground to use DI component

timheuer committed:
Changed rest endpoint name
- changed to 'rest' as named endpoint
- added to reference as a ConnectionString_{name}_rest (for support for SK)
- fixed tests

davidfowl commented on playground/Qdrant/Qdrant.ServiceDefaults/Extensions.cs:
```diff
@@ -0,0 +1,126 @@
+using Microsoft.AspNetCore.Builder;
```
The playground samples share a service defaults.

timheuer replied:
Fixed in next update

davidfowl commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -0,0 +1,106 @@
+        builder.WithEnvironment(context =>
+        {
+            context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}"] = qdrantResource.Resource.ConnectionStringExpression;
+            context.EnvironmentVariables[$"Parameters__{qdrantResource.Resource.ApiKeyParameter.Name}"] = qdrantResource.Resource.ApiKeyParameter.Value;
```
Why `Parameters__`? This should be written into a component specific config section. Also you should directly set the parameter, not the parameter value.

e.g.:

```
```suggestion
            context.EnvironmentVariables[$"Qdrant__ApiKey"] = qdrantResource.Resource.ApiKeyParameter;
```

timheuer replied:
Certainly `Qdrant__XXXX` is better.  For component-specific sectio though, since this is named resource value, were you thinking this would actually be `Aspire__Qdrant__[name]__ApiKey`?

timheuer replied:
Actually, looking at other components, is this correct? Looking at MySQL for example, there would be one Aspire:MySqlConnectorSettings:ConnectionString value -- if I have 2 MySQL resources, this would be shared between them as the component config is not per-named resource...correct? /cc @eerhardt

davidfowl commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -0,0 +1,106 @@
+    {
+        builder.WithEnvironment(context =>
+        {
+            context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}"] = qdrantResource.Resource.ConnectionStringExpression;
```
```suggestion
            context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}"] = new ConnectionStringReference(qdrantResource.Resource, optional: false);
```

timheuer replied:
Fixed in next update

davidfowl commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -0,0 +1,106 @@
+            context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}"] = qdrantResource.Resource.ConnectionStringExpression;
+            context.EnvironmentVariables[$"Parameters__{qdrantResource.Resource.ApiKeyParameter.Name}"] = qdrantResource.Resource.ApiKeyParameter.Value;
+
+            var restEndpointUrl = qdrantResource.Resource.GetEndpoint("rest")?.Url;
+            if (restEndpointUrl != null)
+            {
+                context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}_rest"] = restEndpointUrl;
+            }
```
We're missing an API here, ideally this would be:

```suggestion
            var restEndpointUrl = qdrantResource.Resource.GetEndpoint("rest");
            if (restEndpointUrl.Exists)
            {
                context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}_rest"] = restEndpointUrl;
            }
```

davidfowl commented on src/Components/Aspire.Qdrant/AspireQdrantExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+            settings.ConnectionString = connectionString;
+        }
+
+        if (builder.Configuration[$"Parameters:{connectionName}-ApiKey"] is string apiKey)
+        {
+            settings.ApiKey = apiKey;
+        }
```
This component needs a schema, it shouldn't use parameters, but instaed it should use something like `Qdrant:{connectionName}:ApiKey`

timheuer replied:
@davidfowl it has a schema, see my comment on the other PR feedback as well -- agreed a better named one is better, but the component config schema's don't account for per-named resources (and their typed settings class)

eerhardt replied:
2 thoughts here:

1. For other components that look like this (need a URL and an API Key), we've defined a connection string format ourselves. See 
https://github.com/dotnet/aspire/blob/0d3d68e0d59c1e45d36484761f1516bcfc6e1bbd/src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs#L50-L72

https://github.com/dotnet/aspire/blob/0d3d68e0d59c1e45d36484761f1516bcfc6e1bbd/src/Components/Aspire.Azure.Search.Documents/AzureSearchSettings.cs#L58-L80

I'm not sure we should invent yet another pattern for this component.

2. IMO, the correct fix/design here is https://github.com/dotnet/aspire/issues/1765.

timheuer replied:
What is the suggested change here? In this case the ApiKey is not a part of the connection string, it's passed as an arg to the client lib.

eerhardt replied:
> In this case the ApiKey is not a part of the connection string, it's passed as an arg to the client lib.

That is no different than the 2 scenarios I linked to above. Here's where the ApiKeys get passed as an arg to the client lib for them:

https://github.com/dotnet/aspire/blob/126b1855ba08d7670747d59d2a8048d36d1c1c99/src/Components/Aspire.Azure.AI.OpenAI/AspireAzureOpenAIExtensions.cs#L118-L119

https://github.com/dotnet/aspire/blob/126b1855ba08d7670747d59d2a8048d36d1c1c99/src/Components/Aspire.Azure.Search.Documents/AspireAzureSearchExtensions.cs#L114

timheuer replied:
Should IConnectionStrings move more generically from Azure.Component?

eerhardt replied:
> Should IConnectionStrings move more generically from Azure.Component?

No, the only reason it exists is because the base `AzureComponent` needs to call it on the Settings object. In your case, the `AspireQdrantExtensions` class can just call an internal method on the `QdrantSettings` class.

timheuer replied:
Fixed in latest update

timheuer committed:
Addressing PR feedback
- Changed parameters to component config
- change playground app to keyed services
- changed playground to use shared servicedefaults

roji commented:
Really nice to see this!

> The QdrantClient for .NET uses gRPC by default but Semantic Kernel does not use that client library so the other exposed endpoint is helpful.

FWIW reimplementing the SK connector over the SDK is something that's on my TODO list... that would make the SK connector more efficient thanks to gRPC etc.

> The component doesn't seem to be rich in trace/metrics so those are not implemented...ideas?

Can you point out what trace/metrics in e.g. the Python client look like? This may be something we can improve.

timheuer committed:
Fixing volume mounts to correct target location

timheuer committed:
Add missing README to component

timheuer committed:
Adding logo usage comment to readme after permission from Andrey V from Qdrant

timheuer committed:
Changed playground sample
- matching sample Qdrant/.NET workbook sample

timheuer commented:
> FWIW reimplementing the SK connector over the SDK is something that's on my TODO list... that would make the SK connector more efficient thanks to gRPC etc.

This is great to hear @roji! For now the Resource will expose both endpoints so the SK SDK as-is can get the `ConnectionString` REST endpoint to use in the connector.

timheuer committed:
PR Feedback: Name maps to client name used

timheuer commented on src/Components/Aspire.Qdrant/Aspire.Qdrant.csproj:
```diff
@@ -0,0 +1,25 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Qdrant.Client" />
```
Fixed, plumed through to config

eerhardt commented on src/Aspire.Hosting.Qdrant/QdrantServerResource.cs:
```diff
@@ -0,0 +1,42 @@
+    /// </summary>
+    public ReferenceExpression ConnectionStringExpression =>
+       ReferenceExpression.Create(
+            $"http://{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
```
> http://

This seems wrong. I don't think we should be hard-coding the scheme in the Expression.

davidfowl replied:
that's right, the scheme is part of the endpoint reference:

PrimaryEndpoint.Property(EndpointProperty.Scheme)

timheuer replied:
@eerhardt How do we handle resources that don't have a TLS certificate by default? That's the situation here. https://qdrant.tech/documentation/guides/security/#tls

davidfowl replied:
We can't do https here but it should use the scheme defined by the endpoint.

timheuer replied:
changed to scheme already in latest PR feedback.  As an aside, SSL could be enabled using dotnet dev-certs but felt hacky (but got it working).  Anyhow, PR update to use scheme

timheuer committed:
PR feedback on connection sring
- Moved to Endpoint/Key connection string
- Renamed to Key
- Modified schema to match/updated tests
- Changed primary endpoint to use scheme instead of hardcode
- removed env var for API key only
- fixed renamed component proj location in sln

timheuer committed:
Updating readme/comments to match config

davidfowl commented on src/Aspire.Hosting.Qdrant/QdrantServerResource.cs:
```diff
@@ -0,0 +1,42 @@
+    /// </summary>
+    public ReferenceExpression ConnectionStringExpression =>
+       ReferenceExpression.Create(
+            $"Endpoint={PrimaryEndpoint.Scheme}://{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)};Key={ApiKeyParameter}");
```
```suggestion
            $"Endpoint={PrimaryEndpoint.Property(EndpointProperty.Scheme)}://{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)};Key={ApiKeyParameter}");
```

timheuer committed:
Endpoint property typo fix

mitchdenny commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -0,0 +1,98 @@
+            context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}"] = new ConnectionStringReference(qdrantResource.Resource, optional: false);
+
+            var restEndpointUrl = $"Endpoint={qdrantResource.Resource.GetEndpoint("rest").Url};Key={qdrantResource.Resource.ApiKeyParameter.Value}";
+            if (restEndpointUrl is not null)
```
This will never be null?

davidfowl replied:
yeah we need a new API for this.

timheuer replied:
I realize I'm checking the wrong thing for null, but is it really that `.Exists()` is just going to be a null check anyway?

```
var restEndpoint = qdrantResource.Resource.GetEndpoint("rest");
if (restEndpoint is not null) //Exists?
{
    context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}_rest"] = $"Endpoint={restEndpoint.Url};Key={qdrantResource.Resource.ApiKeyParameter.Value}";
}
```

or more a `TryGetEndpoint`?

davidfowl replied:
it would make sure there's a backing endpoint annotation. Since this change is in main, maybe we just expose it in this PR.

timheuer replied:
It's possible for an `EndpointReference` on a `Resource` to not have annotations?!

davidfowl replied:
The EndpointReference is a promise so it never returns null. You get the reference and asks questions about it 😄 .

timheuer replied:
Wait, so is it already there? `IsAllocated` seems to validate the annotation exists...

davidfowl replied:
Right that's if the endpoint is allocated, I guess you can use that for this scenario. Exists is different. Exists says do I have the EndpointAnnotation. IsAllocated tells you if the endpoint is allocated by DCP. In publish mode, none of the endpoints are allocated, but they all exist.

davidfowl replied:
Wait is this endpoint ever not defined?

timheuer replied:
No it always is defined.  The resource is created always with 2 endpoints, primary and this named one.
```
return builder.AddResource(qdrant)
    .WithImage(QdrantContainerImageTags.Image, QdrantContainerImageTags.Tag)
    .WithHttpEndpoint(hostPort: port, containerPort: QdrantPortGrpc, name: QdrantServerResource.PrimaryEndpointName)
    .WithHttpEndpoint(hostPort: port, containerPort: QdrantPortHttp, name: "rest")

// other parts snipped
```

Maybe that makes you feel better about the simpler check?

But if not, is this what you had in mind for the Exists? (virtually same as Allocated, just not checking the Allocated check?

```
public bool Exists => GetAnnotation() is not null;

private EndpointAnnotation GetAnnotation()
{
    var endpoint = _endpointAnnotation ??= Resource.Annotations.OfType<EndpointAnnotation>().SingleOrDefault(a => StringComparers.EndpointAnnotationName.Equals(a.Name, EndpointName));
    ArgumentNullException.ThrowIfNull(endpoint, $"The endpoint `{EndpointName}` does not exist for the resource `{Resource.Name}`.");
    return endpoint;
}
```

timheuer committed:
PR Feedback
- No need for endpoint null check
- Fixed tests

davidfowl commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -84,13 +84,12 @@ public static IResourceBuilder<ProjectResource> WithReference(this IResourceBuil
-            }
+            // REST endpoint
+            var restEndpoint = qdrantResource.Resource.GetEndpoint("rest");
+            context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}_rest"] = $"Endpoint={restEndpoint?.Url};Key={qdrantResource.Resource.ApiKeyParameter}";
```
```suggestion
            context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}_rest"] = $"Endpoint={restEndpoint};Key={qdrantResource.Resource.ApiKeyParameter}";
```

davidfowl replied:
My bad! 
```suggestion
            context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}_rest"] = ReferenceExpression.Create($"Endpoint={restEndpoint?.Url};Key={qdrantResource.Resource.ApiKeyParameter}");
```

timheuer committed:
Cleanup of ServiceDefaults
Change to ReferenceExpression.Create

davidfowl commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -0,0 +1,97 @@
+
+            // REST endpoint
+            var restEndpoint = qdrantResource.Resource.GetEndpoint("rest");
+            context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}_rest"] = ReferenceExpression.Create($"Endpoint={restEndpoint};Key={qdrantResource.Resource.ApiKeyParameter}");
```
One way to make this a bit cleaner is to make it a property on the resource and just use it here.

timheuer replied:
Like this?

Resource:
    public ReferenceExpression RestConnectionStringExpression =>
        ReferenceExpression.Create(
            $"Endpoint={PrimaryEndpoint.Property(EndpointProperty.Scheme)}://{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}_rest;Key={ApiKeyParameter}");

WithReference overload:
// primary endpoint (gRPC)
context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}"] = qdrantResource.Resource.ConnectionStringExpression;

// REST endpoint
context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}_rest"] = qdrantResource.Resource.RestConnectionStringExpression;

eerhardt commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -0,0 +1,97 @@
+
+            // REST endpoint
+            var restEndpoint = qdrantResource.Resource.GetEndpoint("rest");
+            context.EnvironmentVariables[$"ConnectionStrings__{qdrantResource.Resource.Name}_rest"] = ReferenceExpression.Create($"Endpoint={restEndpoint};Key={qdrantResource.Resource.ApiKeyParameter}");
```
Who reads this `_rest` environment variable?

timheuer replied:
As noted in the original PR, Qdrant exposes two endpoints for access: gRPC (which the client component uses by default), and a REST endpoint (btw, I'm using their terms).  The second is currently what you'd use to compose with SK (although Shay is hoping to change that but not there yet) *and also* is where the Qdrant management web UI is exposed (note: this is disabled in publish mode).

eerhardt commented on tests/Aspire.Hosting.Tests/Qdrant/AddQdrantTests.cs:
```diff
@@ -0,0 +1,220 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
We should also have component tests. See the other components for examples.

It doesn't look like there is a Qdrant test container (yet). https://github.com/testcontainers/testcontainers-dotnet/issues/992. But we can make the tests before that, and just have them check if a connection can be made and skip if it can't. Then you can run the tests with a local Qdrant container running.

roji replied:
Yeah, FYI I was in touch recently with @russcam about reviving that PR, hopefully he'll have a bit of time soon.

russcam replied:
I've updated the PR and requested the maintainers to reopen it (I don't have ability to reopen it- guessing there's some repo setting preventing this).

eerhardt commented on src/Components/Aspire.Qdrant/README.md:
```diff
@@ -0,0 +1,107 @@
+# Aspire.Qdrant.Client library
```
Can you also add entries for this component in:

* https://github.com/dotnet/aspire/blob/main/src/Components/Aspire_Components_Progress.md
* https://github.com/dotnet/aspire/blob/main/src/Components/Telemetry.md

eerhardt commented on src/Components/Aspire.Qdrant/AspireQdrantExtensions.cs:
```diff
@@ -0,0 +1,97 @@
+        {
+            if (settings.Endpoint is not null)
+            {
+                return new QdrantClient(settings.Endpoint, apiKey: settings.Key);
```
Is there any sort of logging, tracing, metrics, health checks that can be enabled?

timheuer replied:
On the client, not yet. I was looking to actually wire up the metrics for the server _via_ the client, but not doing that ;-).

Shay is hopefully going to get some time to work on adding Tracing but it's just not in yet. Tracking: https://github.com/qdrant/qdrant-dotnet/issues/49

eerhardt replied:
It looks like Logging is supported.

https://github.com/qdrant/qdrant-dotnet/blob/b3b80fddbb82bc36c9de9ea7100da93b3624c742/src/Qdrant.Client/QdrantClient.cs#L59-L63

davidfowl commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -0,0 +1,97 @@
+    /// <param name="builder">An <see cref="IResourceBuilder{T}"/> for <see cref="ProjectResource"/></param>
+    /// <param name="qdrantResource">The Qdrant server resource</param>
+    /// <returns></returns>
+    public static IResourceBuilder<ProjectResource> WithReference(this IResourceBuilder<ProjectResource> builder, IResourceBuilder<QdrantServerResource> qdrantResource)
```
Good catch!

timheuer replied:
Yep, will change -- FYI I noticed this is in AWS Hosting package a well.

timheuer commented on src/Components/Aspire.Qdrant/AspireQdrantExtensions.cs:
```diff
@@ -0,0 +1,97 @@
+        {
+            settings.ParseConnectionString(connectionString);
+        }
+
+        if (builder.Configuration[$"{DefaultConfigSectionName}:{connectionName}:Key"] is string apiKey)
+        {
+            settings.Key = apiKey;
+        }
```
What if I want to override the key value.

timheuer replied:
ahh, i see what you mean, you aren't saying remove it from a config, just remove this logic here.  If so, ack.

eerhardt replied:
Yep. just those 5 lines of code can be deleted. The rest looks right to me.

timheuer committed:
More PR Feedback
- Rename component correctly
- Add component client tests
- Add component logging (and document)
- Add property for 2nd endpoint on resource directly

timheuer commented:
@eerhardt I think I addressed most of your feedback (thanks for the pointers on logging).

mitchdenny commented:
On the hosting side this is looking pretty good. I'm wondering whether when we deploy to ACA we should favor the use of the dev services (it supports quadrant). If that is the case then perhaps there should be an Aspire.Hosting.Azure.Qdrant package?

davidfowl commented:
We don’t favor any of the aca services anymore so I would day no. Not as part of this PR and we’d do something like redis first (which requires container app cdk support).

Might be worth spiking PublishAsAcaRedis or something similar

glennc commented on src/Shared/QdrantLogo_256x.png:
Are there any permission gathering we need to do in order to include the logo? Are we inline with their usage guides?

timheuer replied:
Yes already did -- put the disclaimer of use on the README

eerhardt committed:
PR feedback

eerhardt reviewed (commented):
This is looking really great. Thanks for the contribution @timheuer!

I pushed some minor changes and fix ups to the PR.

If you are ready to take this out of "Draft", I think it can be approved. The one remaining comment I have can be addressed in a follow up PR if we think we should do it.

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into qdrant

eerhardt committed:
Fix up the playground to run with latest changes.

eerhardt committed:
Rename QdrantSettings to QdrantClientSettings to match the pattern in other components.

Add class level summary docs.

timheuer commented on playground/Qdrant/Qdrant.ApiService/Program.cs:
```diff
@@ -0,0 +1,95 @@
+            Vectors = new [] {0.106827796f, 0.21676421f},
+            Payload =
+            {
+                ["title"] = "Pulp Function"
```
From a sample #shrug

eerhardt reviewed (approved):
Thanks again for the contribution!

davidfowl commented on src/Aspire.Hosting.Qdrant/QdrantServerResource.cs:
```diff
@@ -0,0 +1,50 @@
+    /// </summary>
+    public ReferenceExpression RestConnectionStringExpression =>
+        ReferenceExpression.Create(
+            $"Endpoint={PrimaryEndpoint.Property(EndpointProperty.Scheme)}://{PrimaryEndpoint.Property(EndpointProperty.Host)}:6333;Key={ApiKeyParameter}");
```
Why is this port hard coded?

timheuer replied:
Converting this feedback to a bug since this PR merged FYI @eerhardt

eerhardt replied:
Resolved this with https://github.com/dotnet/aspire/pull/3422

davidfowl commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+        var qdrant = new QdrantServerResource(name, apiKeyParameter);
+        return builder.AddResource(qdrant)
+            .WithImage(QdrantContainerImageTags.Image, QdrantContainerImageTags.Tag)
+            .WithHttpEndpoint(port: port, targetPort: QdrantPortGrpc, name: QdrantServerResource.PrimaryEndpointName)
```
Using the same port for grpc and http is not correct. Specifying a port will clash.

eerhardt replied:
Resolved this with https://github.com/dotnet/aspire/pull/3422
